### ISSUE TITLE
v0.12 Discussion

### ISSUE THREAD

**Issue by jtopjian** [terraform-providers/terraform-provider-openstack#548]
This issue is meant to be a central area to discuss Terraform v0.12 related items.

### Official Status: Should be Compatible

At this time, all acceptance tests are passing with v0.12.

If you have begun testing this provider with v0.12 and have run into problems, please let us know - reports will help prepare us for full v0.12 support.

### Background

Terraform v0.12 is meant to be a major upgrade which includes _a lot_ of new features. The majority of users will be able to upgrade without having to make any changes. However, in order to reach that point, we need to test the OpenStack provider to make sure it's fully compatible.

Terraform v0.12 has not been released yet. You can download an alpha version of it from [here](https://releases.hashicorp.com/terraform/).

All provider plugins _must_ be recompiled with the Terraform core v0.12 code in order to work with v0.12. If you want to test the OpenStack provider with v0.12, you will have to compile it from source, which includes doing a vendor upgrade to `github.com/hashicorp/terraform`.

We will have a binary release of the OpenStack provider made available once v0.12 is released and all testing has passed. This release will be available by using the usual methods of either `terraform init` or downloading and installing the binary yourself from https://releases.hashicorp.com/terraform-provider-openstack/.

More information about v0.12 can be read about here:

* https://www.terraform.io/upgrade-guides/0-12.html
* https://www.hashicorp.com/blog/terraform-0-1-2-preview

### Known Issues

* Unchecked boxes indicate an issue has been identified.
* Checked boxes indicate the issue was fixed.

- [x] Issues with reading in environment variables. ([link](https://github.com/terraform-providers/terraform-provider-openstack/issues/548#issuecomment-448462322))
- [x] Test fixtures and documentation need to reflect new syntax for declaring arguments/attributes of TypeMap ([link](https://github.com/terraform-providers/terraform-provider-openstack/pull/560))
- [x] Change `TypeString` to `TypeBool` for `admin_state_up` and `shared` arguments of the `openstack_networking_network_v2`. ([link](https://github.com/terraform-providers/terraform-provider-openstack/pull/578))
- [x] Issues with `d.GetOkExists` ([link](https://github.com/hashicorp/terraform/issues/21007))

### How to Test

:warning:  __Do not run this on your production configurations or anything you have already deployed! You will not be able to run Terraform v0.11 afterwards!__ :warning: 

> These instructions _will_ change as v0.12 development progresses!

These instructions assume you have a Go development environment set up and are familiar with compiling Go applications from source.

You will need to use Go 1.11.5 or later with `GO111MODULE=on`.

First, download a Terraform alpha release and store it in a temporary directory:

```shell
$ mkdir t12-test
$ cd t12-test
$ https://releases.hashicorp.com/terraform/0.12.0-rc1/terraform_0.12.0-rc1_linux_amd64.zip
$ unzip terraform_0.12.0-rc1_linux_amd64.zip
```

Next, download the `terraform-provider-openstack` source:

```shell
$ git clone https://github.com/terraform-providers/terraform-provider-openstack
$ cd terraform-provider-openstack
$ git checkout -b v012-testing
```

At this point, you will be on a clean branch and ready to update the vendor dependencies. You should use a clean branch each time you do this - don't try to re-use a previous branch/attempt.

```shell
$ make build
```

Finally, move the compiled binary to your `t12-test` directory:

```shell
$ mv $GOPATH/bin/terraform-provider-openstack ~/t12-test
```

You can now test your configurations by running `terraform` by calling the full path `~/t12-test/terraform` instead of `terraform` as normal.

**Comment by jtopjian** [comment]
While writing the above instructions, I ran into an issue where the environment variables `OS_FLAVOR_NAME` and `OS_FLAVOR_ID` are not being recognized in `openstack_compute_instance_v2`. It's possible this is because `flavor_name` and `flavor_id` are also computed fields - this might be a bug or might be a change we will need to account for.

One possible solution is by not using `schema.EnvDefaultFunc` for `flavor_name` and `flavor_id`. We already do this for `image_name` and `image_id` for _some_ reason (that was 4 years ago - I can't remember why).

**Comment by ozerovandrei** [comment]
@jtopjian can we prepare special OpenLab testing suite to just run all acceptance tests and see if they're failing?

**Comment by kayrus** [comment]
Terraform v0.12 supports validation.IntInSlice https://github.com/hashicorp/terraform/blob/v0.12.0-alpha4/helper/validation/validation.go#L108
Could be used to check IP version.

**Comment by dleske** [comment]
I may be missing something in my environment but I was getting API errors on Terraform trying to use the provider:```$ t12-20190326/terraform plan

Error: Failed to instantiate provider "openstack" to obtain schema: Incompatible API version with plugin. Plugin version: 4, Client versions: [5]```

...until I updated `go.mod` to specify the Terraform core version as `github.com/hashicorp/terraform v0.12.0-beta1` (it was at v0.11.12-beta1 or something, I'm not sure something wasn't munging them together).  Then I was able to build and when I last left it I was able to generate a plan and start working with TF v0.12 and OpenStack objects, though I haven't yet actually tried deploying anything.

Thanks for all your work on this!

**Comment by jtopjian** [comment]
There appears to be some test failures due to `d.GetOkExists`: https://github.com/hashicorp/terraform/issues/21007

**Comment by jtopjian** [comment]
Terraform v0.12-rc1 support has been merged into master. You can compile a binary and use it with either v0.11 or v.12.

Please let us know if anyone runs into any bugs.

**Comment by jtopjian** [comment]
Version 1.19.0 has been released, which includes support for Terraform v0.12. I'm going to close this issue. If any issues are discovered, please open a separate bug report.
